sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -318,16 +333,32 @@ class SpillableHostBufferHandle private (
     materialized
   }
 
+  private var toSpill: Option[HostMemoryBuffer] = None
+
   override def spill(): Long = {
     if (!spillable) {
       0L
     } else {
-      val spilled = synchronized {
-        if (disk.isEmpty && host.isDefined) {
+      val thisThreadSpills = synchronized {
+        if (disk.isEmpty && host.isDefined && toSpill.isEmpty) {
+          toSpill = host
+          // incRefCount here so that if close() is called
+          // while we are spilling, we will prevent the buffer being freed
+          toSpill.get.incRefCount()
+          true
+        } else {
+          false
+        }
+      }
+      val spilled = if (thisThreadSpills) {
+        val buf = toSpill.get
+        withResource(buf) { _ =>
```

Comment:
Why not

```
withResouce(toSpill.get) { buf =>
```

It just makes it so we cannot reference buf outside of the range that it is valid in.